### ISSUE TITLE
Type checking only declarations should be undefined in runtime context

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1010]
**Describe the bug**
Pyright allows symbols defined under an `if TYPE_CHECKING` block, which are undefined at runtime, to be left unquoted outside of delayed annotations.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That Pyright would report that `foo` is undefined on lines 8 and 10.

**Screenshots or Code**
```python
# pyright: strict

from typing import TYPE_CHECKING, cast

if TYPE_CHECKING:
    foo = str

bar = cast(foo, '')     # NameError

def baz(value: foo):    # Also a NameError
    pass
```

**VS Code extension or command-line**
Extension, 1.1.68

**Comment by erictraut** [comment]
This is the intended behavior and is consistent with other type checkers like mypy.

It is legitimate (and quite typical) to use the following construct:
```
if TYPE_CHECKING:
   foo = str
else:
   foo = Any
```

A type checker cannot assume that `TYPE_CHECKING` is both True and False. PEP 484 is clear that type checkers must assume that `TYPE_CHECKING` is True and analyze the file accordingly. It cannot analyze the code in the `else` statement above, for example, because it has been told not to, and doing so will likely generate errors.

I understand that the use of `TYPE_CHECKING` can lead to runtime errors that are not caught by the static analyzer. We generally try to avoid using it in our code because of this. There are cases where it's necessary though. You need to use it with caution.

**Comment by layday** [comment]
Well, that seems rather ill-conceived.  I had not considered that using an `else` to overwrite a symbol defined in `TYPE_CHECKING` would be legitimate; but that `TYPE_CHECKING` would be used to define type aliases without incurring a runtime cost.  Apparently there are classes in the stdlib which can be subscripted when type checking but not at runtime, like `queue.Queue`, which I didn't realise either.  Anyhow, thanks for explaining.
